### ISSUE TITLE
Remove dependency on github.com/docker/docker

### ISSUE THREAD

**Issue by peter-edge** [fsouza/go-dockerclient#260]
One of the biggest pains for me with go-dockerclient is the dependency on github.com/docker/docker. The only thing that is used is the tar functionality in https://github.com/docker/docker/tree/master/pkg/archive, and a couple of file utilities (Matches, Rel) that can be easily redone.

The problem is, docker's tar implementation is probably the best I know of in the golang world. So we don't want to lose that. Also, the archive package depends on a lot of sub-packages:

github.com/docker/docker/pkg/archive
github.com/docker/docker/pkg/fileutils
github.com/docker/docker/pkg/ioutils
github.com/docker/docker/pkg/pools
github.com/docker/docker/pkg/promise
github.com/docker/docker/pkg/system
github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar

However, having to download docker/docker each time I want to make a new container is a real big pain.

Since this repo already depends on the docker license, could we literally vendor out that code and put it in a sub-package here? I don't think docker will ever split up their repo, so that might be the only way. The unfortunate part is that we lose changes upstream to pkg/, but I am almost willing to have that happen. We could have a script in here that basically does:

```
go get -u github.com/docker/docker
cp -R $GOPATH/github.com/docker/docker/pkg/* vendor/docker/pkg/
cp -R $GOPATH/github.com/docker/docker/vendor/src/code.google.com/p/go/src/pkg/archive/tar/ ... # you get the idea
```

Just run it every so often to selectively get updates.

Maybe there is a better way to do this, but whenever I'm running my code inside a Docker container, and my code relies on go-dockerclient and I need to get all the dependencies, there's a huge wait to get github.com/docker/docker. It's just no fun, and all just for some tar functionality.

**Comment by fgrehm** [comment]
Not sure if you came across that yet but here's some background on it: GH-178

**Comment by peter-edge** [comment]
Ah shoot.

The primary reason this is such a pain is when you're building new Docker images with this, library, ironically. Would you be open to a script that just did a selective pre-commit copy, like above?

**Comment by fsouza** [comment]
@peter-edge how often do you need to build a new Docker image? What's your processing of doing it?

Sorry for the delay!

**Comment by archisgore** [comment]
+1 on this. This adds resource cost to build, download and even godep save to be pulling on ALL of docker and building it every time.

**Comment by peter-edge** [comment]
I build a lot of images for testing, `go get` is not exactly the world's greatest command, and about 50% of the time downloading dependencies is spent on github.com/docker/docker, just because of this requirement. I'm sure I'm not the only one who has this issue :) See Use https://github.com/kardianos/vendor for external packages #277 for an extremely clean solution (imo) :)

**Comment by peter-edge** [comment]
Closing 277, use https://github.com/fsouza/go-dockerclient/pull/278 instead.

**Comment by fsouza** [comment]
Closed by #278. Thanks @peter-edge.
